src/main/java/frc/robot/subsystems/vision/VisionSubsystem.java

```diff
@@ -1,175 +1,306 @@
+   * calculated the robot's pose. It adds the pipeline latency, capture latency, and json parsing
+   * latency.
+   */
+  public double getLatencySeconds(int limelightNumber) {
+    return (limelightEstimates[limelightNumber].latency) / 1000.0;
+  }
 
+  /**
+   * Gets the average distance between the specified limelight and the April Tags it sees
+   *
+   * @param limelightNumber the number of the limelight
+   * @return the average distance between the robot and the April Tag(s) in meters
+   */
+  public double getLimelightAprilTagDistance(int limelightNumber) {
+    if (canSeeAprilTags(limelightNumber)) {
+      return limelightEstimates[limelightNumber].avgTagDist;
+    }
     // To be safe returns a big distance from the april tags if it can't see any
     return Double.MAX_VALUE;
   }
 
+  /**
+   * Sets the heading and heading rate of the robot, this is used for deciding between MegaTag 1 and
+   * 2 for pose estimation.
+   *
+   * @param headingDegrees the angle the robot is facing in degrees (0 degrees facing the red
+   *     alliance)
+   * @param headingRateDegrees the rate the robot is rotating, CCW positive
+   */
+  public void setHeadingInfo(double headingDegrees, double headingRateDegrees) {
+    this.headingDegrees = headingDegrees;
+    this.headingRateDegreesPerSecond = headingRateDegrees;
+  }
+
+  /**
+   * Gets the limelight name associated with the specified limelight number/index
+   *
+   * @param limelightNumber the limelight number
+   * @return 0 = limelight-shooter, 1 = limelight-left, 2 = limelight-right
+   */
+  public String getLimelightName(int limelightNumber) {
+    if (limelightNumber == 0) {
+      return VisionConstants.SHOOTER_LIMELIGHT_NAME;
+    } else if (limelightNumber == 1) {
+      return VisionConstants.FRONT_LEFT_LIMELIGHT_NAME;
+    } else if (limelightNumber == 2) {
+      return VisionConstants.FRONT_RIGHT_LIMELIGHT_NAME;
+    }
+    throw new IllegalArgumentException("You entered a number for a non-existent limelight");
+  }
+
+  /** Gets the pose calculated the last time a limelight saw an April Tag */
   public Pose2d getLastSeenPose() {
     return lastSeenPose;
   }
 
-  public String getCurrentlyUsedLimelightName() {
-    return currentlyUsedLimelight;
+  /**
+   * This checks is there is new pose detected by a limelight, and if so, updates the pose estimate
+   *
+   * @param limelightNumber the limelight number
+   */
+  private void checkAndUpdatePose(int limelightNumber) {
+    double last_TX = 0;
+    double last_TY = 0;
+    synchronized (this) { // Synchronize on 'this'
+      double current_TX = LimelightHelpers.getTX(getLimelightName(limelightNumber));
+      double current_TY = LimelightHelpers.getTY(getLimelightName(limelightNumber));
+
+      if (current_TX != last_TX || current_TY != last_TY) {
+        updateLimelightPoseEstimate(limelightNumber);
+        runningThreads.computeIfPresent(limelightNumber, (key, value) -> new AtomicBoolean(true));
```

Comment:
good shit on getting this stuff working, but I'm gonna need u to explain it to me next time I can come to a meeting

Reply:
👍